File: service/app/services/idempotency_service.py
import json
import logging
from typing import Any

import redis.asyncio as aio_redis

from app.cache.constants import RedisKeys
from app.config import Settings

logger = logging.getLogger(__name__)


class IdempotencyService:
    """Двухуровневая дедупликация на базе Redis.

    Уровень 1 (API): предотвращает повторные bulk-send запросы от клиента.
    Уровень 2 (Worker): предотвращает повторную доставку одного уведомления
    при переотправке сообщения из RabbitMQ (at-least-once семантика).
    """

    def __init__(self, redis_client: aio_redis.Redis, settings: Settings) -> None:
        self._redis = redis_client
        self._ttl = settings.idempotency_ttl

    # ── Уровень API ──────────────────────────────────────────────────────────

    async def get_api_response(self, key: str) -> Any | None:
        raw = await self._redis.get(f"{RedisKeys.IDEMPOTENCY_API}{key}")
        if raw is None:
            return None
        logger.debug("Идемпотентность: cache hit для key=%s", key)
        return json.loads(raw)

    async def set_api_response(self, key: str, response: Any) -> None:
        await self._redis.set(
            f"{RedisKeys.IDEMPOTENCY_API}{key}",
            json.dumps(response, default=str),
            ex=self._ttl,
        )

    # ── Уровень Worker ───────────────────────────────────────────────────────

    async def try_claim_worker(self, notification_id: str) -> bool:
        """Атомарно застолбить уведомление для обработки worker-ом.

        Returns:
            True  — этот worker первым заявил на него права.
            False — другой экземпляр уже обработал (дублированная доставка).

        Вторичная защита: основная — атомарный UPDATE статуса в БД
        (NotificationRepository.try_claim_for_processing). Redis — быстрый
        pre-check, позволяющий не трогать БД при очевидных дублях.
        """
        result = await self._redis.set(
            f"{RedisKeys.IDEMPOTENCY_WORKER}{notification_id}",
            "1",
            nx=True,
            ex=self._ttl,
        )
        return result is not None

File: service-laravel/app/Services/IdempotencyService.php
<?php

declare(strict_types=1);

namespace App\Services;

use Predis\Client as Redis;

/**
 * Двухуровневая дедупликация через Redis.
 *
 * Уровень 1 (API):    предотвращает повторные bulk-send запросы от клиента.
 * Уровень 2 (Worker): защищает от повторной доставки at-least-once брокера —
 *                     один notification_id обрабатывается ровно один раз.
 */
final class IdempotencyService
{
    private const API_PREFIX    = 'ik:api:';
    private const WORKER_PREFIX = 'ik:worker:';

    // Значение по умолчанию — 24 часа. Используется в unit-тестах,
    // где сервис создаётся без IoC-контейнера Laravel.
    private const DEFAULT_TTL = 86_400;

    public function __construct(
        private readonly Redis $redis,
        private readonly int   $ttl = self::DEFAULT_TTL,
    ) {}

    // ── Уровень API ───────────────────────────────────────────────────────────

    public function getApiResponse(string $key): ?array
    {
        $raw = $this->redis->get(self::API_PREFIX . $key);

        return $raw !== null ? json_decode($raw, true) : null;
    }

    public function setApiResponse(string $key, array $response): void
    {
        $this->redis->setex(
            self::API_PREFIX . $key,
            $this->ttl,
            json_encode($response, JSON_THROW_ON_ERROR),
        );
    }

    // ── Уровень воркера ───────────────────────────────────────────────────────

    /**
     * Атомарно захватить уведомление для обработки (Redis SET NX).
     *
     * Возвращает true  — этот воркер первый; продолжить доставку.
     * Возвращает false — другой экземпляр воркера уже захватил уведомление
     *                    (сценарий дублирующей доставки at-least-once).
     */
    public function tryClaimWorker(string $notificationId): bool
    {
        $ttl = $this->ttl;

        $result = $this->redis->set(
            self::WORKER_PREFIX . $notificationId,
            '1',
            'EX', $ttl,
            'NX',
        );

        return $result !== null;
    }

    /**
     * Освободить Redis-claim, чтобы следующая попытка могла его захватить заново.
     * Вызывается при перехвате TemporaryProviderException.
     */
    public function releaseWorkerClaim(string $notificationId): void
    {
        $this->redis->del([self::WORKER_PREFIX . $notificationId]);
    }

}
